lxc/storage.go

```diff
@@ -1017,9 +1039,84 @@ func (c *storageCmd) doStoragePoolVolumeShow(client lxd.ContainerServer, pool st
 	return nil
 }
 
+func (c *storageCmd) doStoragePoolVolumeCopy(client lxd.ContainerServer, src string, dst string, rm bool) error {
+	// validate both src and dst string
+	dstVolName, dstVolPool := c.parseVolume("", dst)
+	srcVolName, srcVolPool := c.parseVolume("", src)
+	if dstVolPool == "" || srcVolPool == "" {
+		defaultProfile, _, err := client.GetProfile("default")
+		if err != nil {
+			return fmt.Errorf("No storage pool for source or target volume specified")
+		}
+
+		rootDiskDevice := defaultProfile.Devices["root"]
+		if rootDiskDevice != nil &&
+			rootDiskDevice["pool"] != "" &&
+			rootDiskDevice["type"] == "disk" &&
+			rootDiskDevice["source"] == "" &&
+			rootDiskDevice["path"] == "/" {
+
+			if srcVolPool == "" {
+				srcVolPool = rootDiskDevice["pool"]
+			}
+
+			if dstVolPool == "" {
+				dstVolPool = rootDiskDevice["pool"]
+			}
+		} else {
+			return fmt.Errorf("No storage pool for source or target volume specified")
+		}
+	}
+
+	// Check if the requested storage volume actually exists
+	srcVol, _, err := client.GetStoragePoolVolume(srcVolPool, "custom", srcVolName)
+	if err != nil {
+		return err
+	}
+	client.GetProfile("default")
+
+	if rm && len(srcVol.UsedBy) > 0 {
+		return fmt.Errorf("Storage volume \"%s\" is still in use by profile or containers", srcVol.Name)
+	}
```

Comment:
So for local moves, I think we'd be better off with an API way to do this rather than trying to do it all in the client tool.
Something similar to renames but giving the name of the new pool. This would result in a background operation that will do:
 - Check that the target pool doesn't already contain a volume of that name
 - Ensure no container using this volume are running
 - Update all containers and profiles with the new pool name
 - Copy the volume across
 - Delete the source volume
 - If anything goes wrong, revert the config changes and return the error

This should make it much easier to move a volume from an API point of view and will allow LXD to internally change all needed references.

Reply:
Sounds like you either want a client API extension `MoveStoragePoolVolume` or an argument to `StoragePoolVolumeCopyArgs` because the server needs to know whether it's a copy or a move. Otherwise it can't decide whether to update all containers + profiles with the new pool name. Unless the copy + move case both lead to all containers configs being updated.

Reply:
Yeah, we'd need a MoveStoragePoolVolume for this in the Go API and have that be triggered through POST on a storage volume at the REST API level.

Reply:
So like a rename but with a target pool effectively.

Reply:
Right, but the request still needs to pass an identifier like "copy" or "move" because both copy and post target the same endpoint. Just a new field in the API struct?

Reply:
Hmm, no.

To rename or move: POST /1.0/storage-pools/default/volumes/custom/blah
To copy: POST /1.0/storage-pools/blah/volumes

Reply:
Rather:

To copy: POST /1.0/storage-pools/blah/volumes/custom